File: hive-common/src/main/java/com/amazon/ionhiveserde/objectinspectors/IonUtil.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazon.ionhiveserde.objectinspectors;

import com.amazon.ion.IonValue;

/**
 * Utility methods for Ion values.
 */
public class IonUtil {

    static boolean isIonNull(final IonValue ionValue) {
        return ionValue == null || ionValue.isNullValue();
    }

    public static <T extends IonValue> T handleNull(final T value) {
        return isIonNull(value) ? null : value;
    }
}

File: integration-test/src/test/kotlin/com/amazon/ionhiveserde/integrationtest/IonUtil.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * Ion related utility methods to be used in the test scope.
 */
@file:JvmName("IonUtil")

package com.amazon.ionhiveserde.integrationtest

import com.amazon.ion.*
import com.amazon.ion.system.*
import java.io.File
import java.io.FileOutputStream

/** Standard Ion system with no catalog. */
val DOM_FACTORY: IonSystem = IonSystemBuilder.standard().build()

/** Creates Ion text writer from the local path. */
fun newTextWriterFromPath(path: String): IonWriter =
        IonTextWriterBuilder.standard().build(FileOutputStream(File(path)))

/** Creates Ion binary writer from the local path. */
fun newBinaryWriterFromPath(
        path: String,
        catalog: IonCatalog = SimpleCatalog(),
        vararg imports: SymbolTable): IonWriter {
    return IonBinaryWriterBuilder.standard()
            .withCatalog(catalog)
            .withImports(*imports)
            .build(FileOutputStream(File(path)))
}

// Extensions -------------------------------------------------------------------------------------------------------

/**
 * [IonSystem.singleValue] from a local path.
 */
fun IonSystem.singleValueFromPath(path: String): IonValue = this.singleValue(File(path).readBytes())

/** Struct keys as a sequence. */
fun IonStruct.keys(): Sequence<String> = this.asSequence().map { it.fieldName }

/** Write all values from a string containing an ion document */
fun IonWriter.writeValues(ionDocument: String) =
        IonReaderBuilder.standard().build(ionDocument).use { r -> this.writeValues(r) }
